Content blockers don't hide content in about:blank iframes

b-weinstein opened the pull request:
#### 9b6e3f2ab478195c20afce5b163a9c31f82cdd96
<pre>
Content blockers don&apos;t hide content in about:blank iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=282621">https://bugs.webkit.org/show_bug.cgi?id=282621</a>
<a href="https://rdar.apple.com/134273470">rdar://134273470</a>

Reviewed by Timothy Hatcher and Alex Christensen.

When loading about:blank iframes, we would never consult the ContentExtensionsBackend about if the
load should be blocked, which meant we were never adding the global display none stylesheet to the
frame&apos;s document.

To fix this, explicitly add the global display none stylesheets from the content blockers when
loading an about:blank frame.

* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadSubframe):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
(TEST(ContentRuleList, DisplayNoneInAboutBlankIFrame)): Make sure that the h1 is correctly hidden.
(TEST(ContentRuleList, DisplayNoneInSrcDocIFrame)): Ditto.

Canonical link: <a href="https://commits.webkit.org/286221@main">https://commits.webkit.org/286221@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c39f802961fe771763b66dea2aba8d516993211d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75116 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79572 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26372 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77233 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2340 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58963 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17217 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78183 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49131 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46497 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22024 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24696 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81049 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1509 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67222 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66506 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16540 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10456 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8618 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11607 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2407 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5211 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2432 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3361 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2441 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebCore/loader/SubframeLoader.cpp:
```diff
@@ -328,6 +329,15 @@ RefPtr<LocalFrame> FrameLoader::SubframeLoader::loadSubframe(HTMLFrameOwnerEleme
     frame->protectedLoader()->loadURLIntoChildFrame(url, referrerToUse, subFrame.get());
 
+#if ENABLE(CONTENT_EXTENSIONS)
+    if (url.isAboutBlank() && subFrame->page()) {
```
I am surprised this works with your test, since `srcdoc` uses `about:srcdoc` as the URL. Should this also check `url.isAboutSrcDoc()`?

b-weinstein replied:
I'm happy to add `url.isAboutSrcDoc()` there, but I did verify that the test fails without my change in `SubframeLoader` and passes with it.

b-weinstein replied:
Update: The test was passing with and without the change, but I will keep it around for posterity and to make sure we don't regress this behavior somehow.

xeenon commented on Source/WebCore/loader/SubframeLoader.cpp:
```diff
@@ -328,6 +329,15 @@ RefPtr<LocalFrame> FrameLoader::SubframeLoader::loadSubframe(HTMLFrameOwnerEleme
 
+#if ENABLE(CONTENT_EXTENSIONS)
+    if (url.isAboutBlank() && subFrame->page()) {
+        subFrame->page()->protectedUserContentProvider()->userContentExtensionBackend().forEach([&] (const String& identifier, ContentExtensions::ContentExtension& extension) {
```
Need to use `protectedPage()` or pull this out into a `RefPtr` to use in both places. (The `mac-safer-cpp` will likely complain about this otherwise.)
```suggestion
    RefPtr subFramePage = subFrame->page();
    if ((url.isAboutBlank() || url.isAboutSrcDoc()) && subFramePage) {
        subFramePage->protectedUserContentProvider()->userContentExtensionBackend().forEach([&] (const String& identifier, ContentExtensions::ContentExtension& extension) {
```

xeenon commented on Source/WebCore/loader/SubframeLoader.cpp:
```diff
@@ -328,6 +329,15 @@ RefPtr<LocalFrame> FrameLoader::SubframeLoader::loadSubframe(HTMLFrameOwnerEleme
+#if ENABLE(CONTENT_EXTENSIONS)
+    if (url.isAboutBlank() && subFrame->page()) {
+        subFrame->page()->protectedUserContentProvider()->userContentExtensionBackend().forEach([&] (const String& identifier, ContentExtensions::ContentExtension& extension) {
+            if (auto *styleSheetContents = extension.globalDisplayNoneStyleSheet())
```
```suggestion
            if (RefPtr styleSheetContents = extension.globalDisplayNoneStyleSheet())
```

b-weinstein committed:
Content blockers don't hide content in about:blank iframes
https://bugs.webkit.org/show_bug.cgi?id=282621
rdar://134273470

Reviewed by Timothy Hatcher and Alex Christensen.

When loading about:blank iframes, we would never consult the ContentExtensionsBackend about if the
load should be blocked, which meant we were never adding the global display none stylesheet to the
frame's document.

To fix this, explicitly add the global display none stylesheets from the content blockers when
loading an about:blank frame.

* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadSubframe):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ContentRuleListNotification.mm:
(TEST(ContentRuleList, DisplayNoneInAboutBlankIFrame)): Make sure that the h1 is correctly hidden.
(TEST(ContentRuleList, DisplayNoneInSrcDocIFrame)): Ditto.

Canonical link: https://commits.webkit.org/286221@main
